### DISCUSSION TITLE
Define different fetch instances

### DISCUSSION THREAD

**Discussion by chrissyast** [nuxt/nuxt#19571] [Questions] [answered]
Long-time user of Nuxt 2, struggling to make the transition to v3...

In 2, using Axios, it was possible to define multiple axios instances, e.g.

```
export default function ({ $axios }, inject) {
  const api = $axios.create({
    headers: {
       // headers specific to this endpoint
    }
  })
  api.setBaseURL('http://backendserverurl')
  inject('api', api)

  const thirdPartyApi = $axios.create({
    headers: {
       // headers specific to this endpoint
    }
  })
  thirdPartyApi.setBaseURL('http://thirdPartyApi')
  inject('thirdPartyApi', thirdPartyApi)
};

```

and then call `this.$thirdPartyApi.get('/endpoint')`

but I can't see a way to do that with Nuxt 3/ohmyfetch

**Comment by TheAlexLichter** [comment, marked as answer]
This definitely works. you can use `$fetch.create()`. More info in [the ofetch docs](https://github.com/unjs/ofetch#%EF%B8%8F-create-fetch-with-default-options)

**Reply by chrissyast** [reply]
Thanks for your reply. So, would this replace `useFetch`?

**Reply by TheAlexLichter** [reply]
Depending on your use case. A more appropriate replace would be `useAsyncData` with the newly created fetch instance.

**Reply by chrissyast** [reply]
> Depending on your use case

In previous nuxt 2 apps, I've had a plugin to initialise the separate axios instances, let's say `plugins/axiosInstances.js`, and then I've added that plugin to my `nuxt.config.js`, and then at page/component level whenever I've needed to use an instance it's just `this.$myAxiosInstance.get("/endpoint")`

> useAsyncData with the newly created fetch instance

Do you have an example of this?

**Reply by TheAlexLichter** [reply]
Example is roughly:

```ts
const myFetchInstance = $fetch.create({/**/})
useAsyncData(() => myFetchInstance('URL'))
```

> In previous nuxt 2 apps, I've had a plugin to initialise the separate axios instances, let's say plugins/axiosInstances.js, and then I've added that plugin to my nuxt.config.js, and then at page/component level whenever I've needed to use an instance it's just this.$myAxiosInstance.get("/endpoint")

You can do the same by using the above composable (which you can auto import too) ☺️

**Reply by chrissyast** [reply]
Sorry, I'm getting confused, where is this instance being generated/used in the code? These instances would need to be available within various different pages, so I just want to have each one initialised once and then be able to use them wherever. Also, what happens with the response of useAsyncData?

**Reply by TheAlexLichter** [reply]
You can wrap above lines in an composable, say `/composables/useApi.ts`

```ts
export function useMyAsyncData(url, options = {}){
	return useAsyncData(() => fetch({
      // Static values here
      ...options // override if wanted
    }))
}
```

And then use it in any component/file. `const result = await useMyAsyncData('url')`

**Reply by chrissyast** [reply]
Ok, I think I get it now. Thank you

**Reply by chrissyast** [reply]
Update: I got around to implementing this today and found that the `useAsyncData(() => ...` part can be skipped and

```js
export default async function useMyAsyncData(url: any, options = {}){
  return await $fetch(url, { 
    baseURL: "http://localhost:8080",
    credentials: "include",
    ...options
  })
}
```

was sufficient 👍

**Reply by TheAlexLichter** [reply]
.@chrissyast then it is no composabel anymore but a simple util function. It will also not transfer state from server to client.

**Reply by chrissyast** [reply]
Hmm, I don't really understand what any of that means or if it's something I need to worry about.

**Reply by TheAlexLichter** [reply]
If you use SSR then yes, else no.

TL;DR is - you *want* to use `useAsyncData` as part of your *composable*  if it should be reactive and transfer state from server to client.

**Reply by chrissyast** [reply]
> be reactive and transfer state from server to client

what does this mean?

**Reply by TheAlexLichter** [reply]
Reactive: the `data` value will change as soon as the call is done, params and URL will be watched and the call will be re-executed when they change.

Transfer state from server to client: When you render your site via SSR, the initial request will be handled by the server. The server will do API calls and render the HTML. But somehow, the client has to receive the "state" of the app the server rendered so it can "continue from there". Thus, the state (e.g. all API responses) has to be serialized (commonly to JSON or similar) and sent with the HTML. Nuxt is doing that for you when you use `useState`, `useAsynData`, `useFetch` (or the lazy variants of the last two)

See also https://nuxt.com/docs/getting-started/data-fetching

**Reply by chrissyast** [reply]
If the URL changes you would leave the page anyway, no?

**Reply by TheAlexLichter** [reply]
I meant the URL that should be queried by your composable (as it can be a `ref` or `computed` too).

**Reply by chrissyast** [reply]
Hmm...why would that change? If you change the url then it'll become different data, surely? Like if you've got `const {data: posts} = useAsyncData("/posts")`, you're not going to want to change the url to `("/users")` because then the variable name would be incorrect 🤔

**Reply by TheAlexLichter** [reply]
It is just an option, you don't have to change the URL but it is possible (e.g. manually appending query params or whatnot). ☺️
